README.md
# A3_Regresion_Simbolica
Articulo: Predictor de futuros de tipo de cambio con regresión simbólica y regularización de modelos lineales multivariados.


changelog.txt

# -- --------------------------------------------------------------------------------------------------- -- #
# -- project: Multivariate Linear Regression with Symbolic Regressors and L1L2 Regularization            -- #
# -- for future prices prediction, the case for UsdMxn                                                   -- #
# -- script: changelog.txt : text file with changes specification                                        -- #
# -- author: IFFranciscoME                                                                               -- #
# -- license: GPL-3.0 License                                                                            -- #
# -- repository: https://github.com/IFFranciscoME/A3_Regresion_Simbolica                                 -- #
# -- --------------------------------------------------------------------------------------------------- -- #

Jueves 29.10.2020
- Se agrego archivo changelog para control de cambios
- Cambios en descripcion de proyecto en los docstring de todos los archivos
- Se agrego carpeta "files" para ahi guardar el archivo "simbolic_features.csv"
- Reacomodo de notas en archivo notes.txt
- Cambios menores en formato y comentarios en codigos
- Modificaciones en funcion generadora de features iniciales (producto hadamard y nmax para elegir)
- mas features automaticos (p_nmax = 30)
- 'had_' indicativo para columnas de producto hadamard y '_' para dividir features involucrados
- separar funciones de regresion simple y aquellas con regularizacion
- regresar mas parametros despues de ajuster modelos
- modificados parametros de regreso de regresion simbolica
- Se quito codigo de regresion con regularizacion (esto se va a hacer en R)
- Se quito SymbolicRegression por SymbolicTransformer
- Con la salida de SymbolicTransformer se incluyen los nuevos features en una nueva regresion simple

Viernes 30.10.2020
- Modificaciones a parametros de llamada del SymbolicRegressor.
- Se agrego paper de referencia para leer mas sobre la inicializacion de la profundidad de los arboles.
- Se redactaron "Rutas heuristicas" para elegir parametros de SymbolicTransformer.

- Se hizo el proceso de generacion de variable explicativa desde data.py
- Se separaron los conjuntos de entrenamiento y prueba en data.py

Viernes 22.01.2021
- Modificacion y acomodo de folders del proyecto

Martes 26.01.2021
- Cambiar en data.py los precios historicos por archivo de firstrate con resampleo
- Agregar en main.py y data.py indicativo de separacion 'Train: 2010-2018, Test: 2019, Validate: 2020'
- Cambiar en functions.py: f_features  para calculos de features con volume en lugar de openiterest
  ya que en los nuevos datos no se tiene la columna de openiterest.
- Agregar en functions.py: data_trans para escalamiento y transformacion de datos
- Agregar en data.py lista de parametros para features e hiperparametros para modelo
- Cambiar en functions.py: Nombre de f_features a features
- Quitar en main.py el uso de datos con Quandl
- Cambiar en main.py: el orden del procesamiento y extener los comentarios
- Agregar en functions.py: funcion para regresion (ols y ols-elasticnet)
- Agregar en functions.py: funcion para clasificacion (logistica y logistica-elasticnet)
- Agregar funcionalidad de optimizacion con algoritmos geneticos
- Cambiar en varios scripts que los datos solo sean x_data y y_data en lugar de x_train/y_train x_test/y_test

Miercoles 23.01.2021
- cambiar en functions.py data_trans para hacer transformacion a solo X y no las Y
- cambiar todo en main.py para correr el proceso completo para 4 variaciones de modelos
  ols, ols_en, logistic, logitstic_en sin features symbolicas

- (pendiente) correr codigo para que haya resultados in sample y out of sample para regresion y clasificacion
  utilizando variables sin features symbolicas
- (pendiente) hacer analisis y graficas de resultados para mostrar lo obtenido
- (pendiente) completar codigo para generar features symbolicas 
- (pendiente) correr codigo para que haya resultados in sample y out of sample para regresion y clasificacion
  utilizando variables con features symbolicas
- (pendiente) hacer analisis y graficas de resultados para mostrar lo obtenido
- (pendiente) hacer analisis y graficas de comparaciones entre todo


Domingo 31-01-2020:
-Función para checar heterocedasticidad
-Funcion para visualizar residuales
- Mejor modelo optimizado in sample para regresión con y sin variables simbólicas
- cambio hiperparametros simbolic features
- visualziación de distribuciones variables simbolicas


Interpretacion Jarque-bera
-Interpretación: La prueba de Jarque-Bera utiliza un estadístico de prueba que involucra
la curtosis y la asimetría. Intenta resolver si “Los residuos tienen una distribución normal”,
esta H0 se rechaza si p≤α, o JB≥VC, de lo contrario H0 no se rechaza. (VC de 5%=5.9915)
 -Respuesta: En este caso Pvalue>0.05 y 2.2973<5.9915, por lo que los residuos
 tienen una distribución normal

main.py

"""
# -- --------------------------------------------------------------------------------------------------- -- #
# -- project: Multivariate Linear Regression with Symbolic Regressors and L1L2 Regularization            -- #
# -- for future prices prediction, the case for UsdMxn                                                   -- #
# -- script: main.py : python script with the main functionality                                         -- #
# -- author: IFFranciscoME                                                                               -- #
# -- license: GPL-3.0 License                                                                            -- #
# -- repository: https://github.com/IFFranciscoME/A3_Regresion_Simbolica                                 -- #
# -- --------------------------------------------------------------------------------------------------- -- #
"""

import pandas as pd
import numpy as np
import matplotlib. pyplot as plt
import functions as fn
from data import params, params_reg, all_data
from visualizations import vs
import visualizations as vis

# pd.set_option('display.max_rows', None)                   # sin limite de renglones maximos
# pd.set_option('display.max_columns', None)                # sin limite de columnas maximas
# pd.set_option('display.width', None)                      # sin limite el ancho del display
# pd.set_option('display.expand_frame_repr', False)         # visualizar todas las columnas

# -------------------------------------------------------------------------------------------- Read data -- #
# train data (2011 to 2019) - Use any number of times
#data = ohlc_data['train']

# test data (2019) - Use max 10 times
# data = ohlc_data['train']

# validation data (2020) - Use 1 time at the very end
# data = ohlc_data['val']

# -- ------------------------------------------------------------------------- Exploratory Data Analysis -- #

# Description table
# data.describe()

# OHLC plot
#p_theme = {'color_1': '#ABABAB', 'color_2': '#ABABAB', 'color_3': '#ABABAB', 'font_color_1': '#ABABAB',
 #          'font_size_1': 12, 'font_size_2': 16}
#p_dims = {'width': 1450, 'height': 800}
#p_vlines = [data['timestamp'].head(1), data['timestamp'].tail(1)]
#p_labels = {'title': 'Main title', 'x_title': 'x axis title', 'y_title': 'y axis title'}

# cargar funcion importada desde GIST de visualizaciones
#ohlc = vs['g_ohlc'](p_ohlc=data, p_theme=p_theme, p_dims=p_dims, p_vlines=p_vlines, p_labels=p_labels)

# mostrar plot
# ohlc.show()

# -- ------------------------------------------------------------------------------- Feature Engineering -- #

# Feature engineering (autoregressive and hadamard functions)
data_t, data_y, data_x, features_names = fn.features(p_data=all_data, p_nmax=7)

# Data scaling
data_x = fn.data_trans(p_data=data_x, p_trans='Rubust')

var = fn.variables(data_y, data_x, N=4)
busqueda= fn.busqueda_en_train(var, n=4)
# Rearange of data for regression model

# Rearange of data for classification model
data_cla = {'x_data': data_x, 'y_data': data_y['co_d']}

# -- ---------------------------------------------------------------------------------- Feature analysis -- #

# matriz de correlacion
cor_mat = data_x.corr()

# optimization
#data_op = fn.optimization(p_data=data_cla, p_type='classification', p_params=params,
                          #p_model='logistic', p_iter=500)

# in sample results

# out of sample results

# -- with elastic net regularization

# optimization
#data_op = fn.optimization(p_data=data_cla, p_type='classification', p_params=params,
                          #p_model='logistic_en', p_iter=500)


visualizations.py

"""
# -- --------------------------------------------------------------------------------------------------- -- #
# -- project: Multivariate Linear Regression with Symbolic Regressors and L1L2 Regularization            -- #
# -- for future prices prediction, the case for UsdMxn                                                   -- #
# -- script: visualizations.py : python script with data visualization functions                         -- #
# -- author: IFFranciscoME                                                                               -- #
# -- license: GPL-3.0 License                                                                            -- #
# -- repository: https://github.com/IFFranciscoME/A3_Regresion_Simbolica                                 -- #
# -- --------------------------------------------------------------------------------------------------- -- #
"""

import requests
import plotly.io as pio
import plotly.graph_objects as go
import pandas.plotting as cor
pio.renderers.default = "browser"
import plotly.express as px

# ----------------------------------------------------------------- GIST 1: Data Visualization in Python -- #
base = "https://gist.githubusercontent.com/IFFranciscoME/9b7b842ed368e9118c5ef09bb9b228cf/raw/"
gist = "f39623784704f556a49486bae2e9335f6d6da9e7/gist_p_data_vis"
vs = {}
exec(requests.get(base).content, vs)


def residual(residuales):
    fig = go.Figure()
    fig.add_trace(go.Scatter(x=residuales.index, y=residuales, mode='lines'))
    fig.show()


def correlation(data):
    return cor.scatter_matrix(data, diagonal='kde', alpha=0.8, figsize=(20, 20))

def histograma(data):
    fig = px.histogram(data, x="co", nbins=100)
    fig.show()

data.py

"""
# -- --------------------------------------------------------------------------------------------------- -- #
# -- project: Multivariate Linear Regression with Symbolic Regressors and L1L2 Regularization            -- #
# -- for future prices prediction, the case for UsdMxn                                                   -- #
# -- script: data.py : python script for data collection                                                 -- #
# -- author: IFFranciscoME                                                                               -- #
# -- license: GPL-3.0 License                                                                            -- #
# -- repository: https://github.com/IFFranciscoME/A3_Regresion_Simbolica                                 -- #
# -- --------------------------------------------------------------------------------------------------- -- #
"""

import pandas as pd
import numpy as np
from os import listdir, path
from os.path import isfile, join
import quandl as quandl


# authentication for quandl
quandl.ApiConfig.api_key = "4WLCk2ZUAMAPPyowvAYr"

# -- ------------------------------------------------------------------- Obtain, clean and reformat Data -- #
# -- Historical prices of E-Micro Eur/Usd Futures M6E1 (Front Month)

m6e1 = quandl.get('CHRIS/CME_m6e1', start_date='2018-08-31', end_date='2020-08-31')
m6e1 = m6e1.rename(columns={'Previous Day Open Interest': 'OpenInterest', 'Last': 'Close'})
m6e1 = m6e1[['Open', 'High', 'Low', 'Close', 'OpenInterest', 'Volume']]
m6e1.columns = ['open', 'high', 'low', 'close', 'openinterest', 'volume']
m6e1.reset_index(inplace=True, drop=False)
m6e1 = m6e1.rename(columns={'Date': 'timestamp'}, inplace=False)

# ---------------------------------------------------------------------------- Historical Prices Reading -- #

# the price in the file is expressed as the USD to purchase one MXN
# if is needed to convert to the inverse, the MXN to purchase one USD, uncomment the following line
mode = 'MXN_USD'

# path in order to read files
main_path = 'files/daily/'
abspath_f = path.abspath(main_path)
files_f = sorted([f for f in listdir(abspath_f) if isfile(join(abspath_f, f))])
price_data = {}

# iterative data reading
for file_f in files_f:
    data_f = pd.read_csv(main_path + file_f)
    data_f['timestamp'] = pd.to_datetime(data_f['timestamp'])

    # swap since original is wrong
    high = data_f['high'].copy()
    low = data_f['low'].copy()
    data_f['high'] = low
    data_f['low'] = high

    if mode == 'MXN_USD':
        data_f['open'] = round(1/data_f['open'], 5)
        data_f['high'] = round(1/data_f['high'], 5)
        data_f['low'] = round(1/data_f['low'], 5)
        data_f['close'] = round(1/data_f['close'], 5)

    data_f = data_f[['timestamp', 'open', 'high', 'low', 'close', 'volume']]

    years_f = set([str(datadate.year) for datadate in list(data_f['timestamp'])])
    for year_f in years_f:
        price_data['MP_D_' + year_f] = data_f

# Train data from 2011 to 2018
all_data = pd.concat([price_data[list(price_data.keys())[0]], price_data[list(price_data.keys())[1]],
                      price_data[list(price_data.keys())[2]], price_data[list(price_data.keys())[3]],
                      price_data[list(price_data.keys())[4]], price_data[list(price_data.keys())[5]],
                      price_data[list(price_data.keys())[6]], price_data[list(price_data.keys())[7]],
                      price_data[list(price_data.keys())[8]], price_data[list(price_data.keys())[9]]])


# historical data 
#ohlc_data = {'train1': train_data1, 'train2': train_data2 , 'train3': train_data3, 'test': test_data}

# --------------------------------------------------------------------- Parameters for Symbolic Features -- #
# --------------------------------------------------------------------- -------------------------------- -- #

symbolic_params = {'functions': ["sub", "add", 'inv', 'mul', 'div', 'abs', 'log','max','min'],
                   'population': 2000, 'tournament': 20, 'hof': 20, 'generations': 5, 'n_features':10,
                   'init_depth': (4,12), 'init_method': 'half and half', 'parsimony': 0, 'constants': None,
                   'metric': 'pearson', 'metric_goal': 0.7,
                   'prob_cross': 0.4, 'prob_mutation_subtree': 0.3,
                   'prob_mutation_hoist': 0.1, 'prob_mutation_point': 0.2,
                   'verbose': True, 'parallelization': True, 'warm_start': True}

# ------------------------------------------------------------------------ Hyperparameters for the Model -- #
# ------------------------------------------------------------------------ ----------------------------- -- #

# 100 different values for each parameter
params = {'ratio': np.arange(0, 1, 0.01), 'c': np.arange(0, 2, 0.02)}
params_reg = {'ratio': np.arange(0, 1, 0.01), 'c': np.arange(.2, 1, 0.2)}
